Remove `EntityLink`

kamilmielnik opened the pull request:
Closes #51041

### Description

Removes `EntityLink` & `entity.Link` components to get rid of `EntityObjectLoader` usages.

### How to verify

- Database breadcrumb in `/browse/databases/:id` works
- Link in toast after moving question/model/metric to a different collection works
- Link in toast after moving dashboard to a different collection works

kamilmielnik committed:
Get rid of Database.Link usage in TableBrowser

kamilmielnik committed:
Fix browseDatabase typing

kamilmielnik committed:
Get rid of Collection.Link usage in DashboardMoveModal

kamilmielnik committed:
Get rid of Collection.Link usage in QuestionMoveToast

kamilmielnik committed:
Style links consistently

kamilmielnik committed:
Remove EntityLink

kamilmielnik commented on frontend/src/metabase/dashboard/components/DashboardMoveModal.styled.tsx:
```diff
@@ -1,6 +1,12 @@
 `;
+
+export const CollectionLink = styled(Link)`
+  color: var(--mb-color-brand);
```
I added this color here to make it consistent with Collection link in `QuestionMoveToast`

uladzimirdev replied:
it would be nice if you could use css modules instead

romeovs commented on frontend/src/metabase/questions/components/QuestionMoveToast/QuestionMoveToast.tsx:
```diff
@@ -15,31 +18,24 @@ interface QuestionMoveToastProps {
-      {getMessage(question, collectionLink)}
+      {type === "question" && jt`Question moved to ${collectionLink}`}
+      {type === "model" && jt`Model moved to ${collectionLink}`}
+      {type === "metric" && jt`Metric moved to ${collectionLink}`}
```
Should we handle unknown question types here? Either by showing something like `Item moved to ...` or by throwing (so tests would catch this?

This is one of those things where we might add a type and forget about adding a new message here.

Or we could use `ts-pattern` here so the types catch it.

kamilmielnik replied:
> Should we handle unknown question types here? Either by showing something like Item moved to ... or by throwing (so tests would catch this?

Currently it's not possible to have different card type.

> Or we could use `ts-pattern` here so the types catch it.

If we're going to do this then we should do it everywhere because there's lots of places with similar card type based conditions. It'd deserve a dedicated PR.

ranquild replied:
Use `ts-pattern` with `exhaustive` for a compile-type check

ranquild replied:
No reason not to do it here with ts-pattern now fwiw

kamilmielnik replied:
2b63f7f395365fae8deb3783d8d36b111ed7860e

romeovs commented on frontend/src/metabase/browse/tables/TableBrowser/TableBrowser.jsx:
```diff
@@ -46,15 +42,17 @@ export const TableBrowser = ({
+            databaseCrumb,
             showSchemaInHeader && { title: schemaName },
-          ]}
+          ].filter(Boolean)}
```
When the `databaseCrumb` is loading or errors out, this will show a shorter breadcrumb trail.

<img width="214" alt="Screenshot 2024-12-10 at 18 07 14" src="https://github.com/user-attachments/assets/6fc468cd-798e-45d6-a03a-f31e8de13f5b">
<img width="330" alt="Screenshot 2024-12-10 at 18 07 20" src="https://github.com/user-attachments/assets/b26239f1-f8ec-4175-a40b-4bebad40f4d6">

I think we should show a placeholder instead to avoid confusion.

kamilmielnik replied:
> When the databaseCrumb is loading or errors out, this will show a shorter breadcrumb trail.

d29e5130ef15e290e34f023d27c868944251f042 + e3eff706d3cab17df75d48a6dbc2e0ea26c8de35

I used data from the store as a fallback - now error and loading state are not relevant anymore.
If initial global GET `/api/database` call fails, then app won't work anyway.

> I think we should show a placeholder instead to avoid confusion.

I fully agree but it's out of scope of this PR.

kamilmielnik committed:
Use database from the store if it's already there

kamilmielnik committed:
Remove filter

kamilmielnik committed:
Fix typo

kamilmielnik committed:
Use ts-pattern

uladzimirdev commented on frontend/src/metabase/questions/components/QuestionMoveToast/QuestionMoveToast.tsx:
```diff
@@ -15,31 +19,26 @@ interface QuestionMoveToastProps {
     <ToastRoot>
       <StyledIcon name="collection" />
-      {getMessage(question, collectionLink)}
+      {match(type)
```
nit: looks more like a switch/case for me

kamilmielnik replied:
2b63f7f395365fae8deb3783d8d36b111ed7860e

uladzimirdev reviewed (commented):
When I moved question to another collection, seems, space before collection name is missed

<img width="338" alt="image" src="https://github.com/user-attachments/assets/1724ace6-16e1-4a05-b0d2-a634023d0a9d">

Also I found it weird that moving a dashboard to another collection from the collection view, we show a hint toast which doesn't contain a link to the destination.

kamilmielnik committed:
Bring spacing back

kamilmielnik commented:
> When I moved question to another collection, seems, space before collection name is missed

Good catch, fixed: 41abeae80c16fd69c0b7290e860368342115302f

> Also I found it weird that moving a dashboard to another collection from the collection view, we show a hint toast which doesn't contain a link to the destination.

True, we should probably align this as well.

uladzimirdev committed:
Merge branch 'master' into 51041-entity-link

kamilmielnik committed:
Merge branch 'master' into 51041-entity-link

kamilmielnik committed:
Extract components for links
